### DISCUSSION TITLE
What should be the server specs for Strongswan?

### DISCUSSION THREAD

**Discussion by meteoguzhan** [strongswan/strongswan#1226] [Q&A]
Hello,
I have a ubuntu server with these features and this configuration. How many connections can I open instantly with this server or how many connections can I have without any problem? I had no problem up to 15 or 16 connections but will I have problems with 50 or 100 connections? Thank you in advance for your support and help.

**Features of my server;**
vCPU/s: 1 Intel vCPU
RAM: 1024 MB
Storage: 25 GB NVMe
Bandwidth: 2000 GB

**ipsec version**
```
Linux strongSwan U5.9.5/K5.15.0-43-generic
University of Applied Sciences Rapperswil, Switzerland
See 'ipsec --copyright' for copyright information.
```
**iptables -L**
```
Chain INPUT (policy ACCEPT)
target     prot opt source               destination         
DROP       all  --  anywhere             anywhere             state NEW recent: UPDATE seconds: 300 hit_count: 60 name: DEFAULT side: source mask: 255.255.255.255
           all  --  anywhere             anywhere             state NEW recent: SET name: DEFAULT side: source mask: 255.255.255.255
ACCEPT     all  --  anywhere             anywhere             state RELATED,ESTABLISHED
ACCEPT     all  --  anywhere             anywhere            
DROP       all  --  anywhere             anywhere             state INVALID
ACCEPT     tcp  --  anywhere             anywhere             tcp dpt:ssh
ACCEPT     udp  --  anywhere             anywhere             udp dpt:isakmp
ACCEPT     udp  --  anywhere             anywhere             udp dpt:ipsec-nat-t
DROP       all  --  anywhere             anywhere            

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination         
ACCEPT     all  --  10.101.0.0/16        anywhere             policy match dir in pol ipsec proto esp
ACCEPT     all  --  anywhere             10.101.0.0/16        policy match dir out pol ipsec proto esp
DROP       all  --  anywhere             anywhere
```

**/etc/sysctl.conf**
```
net.ipv4.ip_forward = 1
net.ipv4.ip_no_pmtu_disc = 1
net.ipv4.conf.all.rp_filter = 1
net.ipv4.conf.all.accept_redirects = 0
net.ipv4.conf.all.send_redirects = 0
net.ipv6.conf.all.disable_ipv6 = 1
net.ipv6.conf.default.disable_ipv6 = 1
net.ipv6.conf.lo.disable_ipv6 = 1
net.ipv6.conf.eth0.disable_ipv6 = 1
```

**/etc/ipsec.conf**
```
config setup
  strictcrlpolicy=no
  uniqueids=never

conn roadwarrior
  auto=add
  compress=yes
  aggressive=no
  type=tunnel
  keyexchange=ikev2
  fragmentation=yes
  forceencaps=yes
  mobike=yes

  ike=aes256gcm16-prfsha384-ecp384!
  esp=aes256gcm16-ecp384!

  dpdaction=none
  left=%any
  leftid=@vpn.vpn.vpn
  leftcert=cert.pem
  leftsendcert=always
  leftsubnet=0.0.0.0/0
  right=%any
  rightid=%any
  rightauth=eap-mschapv2
  rightdns=1.1.1.1,1.0.0.1
  rightsourceip=10.101.0.0/16
  rightsendcert=never
```
**/etc/strongswan.conf**
```
charon {
    i_dont_care_about_security_and_use_aggressive_mode_psk = no

    threads = 64
    keep_alive=0
    init_limit_half_open = 500

    load_modular = yes
    plugins {
        include strongswan.d/charon/*.conf
    }
}
include strongswan.d/*.conf
```

**iptables-save**
```
# Generated by iptables-save v1.8.7 on Sun Aug 14 16:25:21 2022
*mangle
:PREROUTING ACCEPT [0:0]
:INPUT ACCEPT [0:0]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [0:0]
:POSTROUTING ACCEPT [0:0]
-A FORWARD -s 10.101.0.0/16 -o enp1s0 -p tcp -m policy --dir in --pol ipsec -m tcp --tcp-flags SYN,RST SYN -m tcpmss --mss 1361:1536 -j TCPMSS --set-mss 1360
COMMIT
# Completed on Sun Aug 14 16:25:21 2022
# Generated by iptables-save v1.8.7 on Sun Aug 14 16:25:21 2022
*filter
:INPUT ACCEPT [0:0]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [5007005:5711223475]
:ufw-after-forward - [0:0]
:ufw-after-input - [0:0]
:ufw-after-logging-forward - [0:0]
:ufw-after-logging-input - [0:0]
:ufw-after-logging-output - [0:0]
:ufw-after-output - [0:0]
:ufw-before-forward - [0:0]
:ufw-before-input - [0:0]
:ufw-before-logging-forward - [0:0]
:ufw-before-logging-input - [0:0]
:ufw-before-logging-output - [0:0]
:ufw-before-output - [0:0]
:ufw-logging-allow - [0:0]
:ufw-logging-deny - [0:0]
:ufw-not-local - [0:0]
:ufw-reject-forward - [0:0]
:ufw-reject-input - [0:0]
:ufw-reject-output - [0:0]
:ufw-skip-to-policy-forward - [0:0]
:ufw-skip-to-policy-input - [0:0]
:ufw-skip-to-policy-output - [0:0]
:ufw-track-forward - [0:0]
:ufw-track-input - [0:0]
:ufw-track-output - [0:0]
:ufw-user-forward - [0:0]
:ufw-user-input - [0:0]
:ufw-user-limit - [0:0]
:ufw-user-limit-accept - [0:0]
:ufw-user-logging-forward - [0:0]
:ufw-user-logging-input - [0:0]
:ufw-user-logging-output - [0:0]
:ufw-user-output - [0:0]
-A INPUT -i enp1s0 -m state --state NEW -m recent --update --seconds 300 --hitcount 60 --name DEFAULT --mask 255.255.255.255 --rsource -j DROP
-A INPUT -i enp1s0 -m state --state NEW -m recent --set --name DEFAULT --mask 255.255.255.255 --rsource 
-A INPUT -m state --state RELATED,ESTABLISHED -j ACCEPT
-A INPUT -i lo -j ACCEPT
-A INPUT -m state --state INVALID -j DROP
-A INPUT -p tcp -m tcp --dport 22 -j ACCEPT
-A INPUT -p udp -m udp --dport 500 -j ACCEPT
-A INPUT -p udp -m udp --dport 4500 -j ACCEPT
-A INPUT -j DROP
-A FORWARD -s 10.101.0.0/16 -m policy --dir in --pol ipsec --proto esp -j ACCEPT
-A FORWARD -d 10.101.0.0/16 -m policy --dir out --pol ipsec --proto esp -j ACCEPT
-A FORWARD -j DROP
COMMIT
# Completed on Sun Aug 14 16:25:21 2022
# Generated by iptables-save v1.8.7 on Sun Aug 14 16:25:21 2022
*nat
:PREROUTING ACCEPT [0:0]
:INPUT ACCEPT [0:0]
:OUTPUT ACCEPT [0:0]
:POSTROUTING ACCEPT [0:0]
-A POSTROUTING -s 10.101.0.0/16 -o enp1s0 -m policy --dir out --pol ipsec -j ACCEPT
-A POSTROUTING -s 10.101.0.0/16 -o enp1s0 -j MASQUERADE
COMMIT
# Completed on Sun Aug 14 16:25:21 2022
```

**Comment by strongX509** [comment]
Several hundred connections shouldn't be a problem.

**Comment by Thermi** [comment]
Entirely depends on what the actual processor is, the load mix, and what your expected throughput is.

**Reply by meteoguzhan** [reply]
> Entirely depends on what the actual processor is, the load mix, and what your expected throughput is.

150 or 200 people will actively use this server for an average of 8 hours. I should not have problems with speed or there should be no interruptions. What do you think is the most logical setting for me? Or if I switch to a higher server with the same settings, will it cause problems? or is this server enough for 150 200 connections and usage.

**Reply by Thermi** [reply]
That doesn't tell us that much. You probably want some way to balance load over different cores with RSS, RPS and RFS. For RSS you need multiple cores first. Otherwise it makes little sense to do RSS. Check if your provider can give you a NIC with as many RX queues as there are cores, or at least two or four, and about that many CPU cores.

Or just throw the speed of the connection and package loss in the wind in high load scenarios (like when people start opening Outlook or doing other things in parallel in the morning)

**Reply by meteoguzhan** [reply]
> That doesn't tell us that much. You probably want some way to balance load over different cores with RSS, RPS and RFS. For RSS you need multiple cores first. Otherwise it makes little sense to do RSS. Check if your provider can give you a NIC with as many RX queues as there are cores, or at least two or four, and about that many CPU cores.
> 
> Or just throw the speed of the connection and package loss in the wind in high load scenarios (like when people start opening Outlook or doing other things in parallel in the morning)

Thank you very much for your reply. Currently, when 15 or 20 people connect, they can get 20 mbps per person. These people do not connect at the same time, they connect at different times, there is no problem with that. I did not encounter a problem when 20 people actively used the server. But this number will increase gradually and will reach 100s and 200s. So I wanted to ask, should I upgrade already?

My maximum server capacity;
vCPU/s:  12 Intel vCPU
RAM: 24576 MB
Storage: 500 GB NVMe
Bandwidth: 12.00 TB

My server I am currently using;
vCPU/s: 1 Intel vCPU
RAM 1024 MB
Storage: 25 GB NVMe
Bandwidth: 2000 GB

Actually, the question I need to ask clearly is that when 200 people use the server at the same time, there will be a continuous download and upload for 24 hours. Should I switch to the server at maximum capacity. What I want is to work in a maximum efficient way.

**Reply by Thermi** [reply]
If you want to increase capacity effectively you MUST be able to use RSS. So verify first if you can use that!

**Reply by meteoguzhan** [reply]
> If you want to increase capacity effectively you MUST be able to use RSS. So verify first if you can use that!

I can use whatever is needed. What is the definition of RSS, I want to use it.

**Reply by meteoguzhan** [reply]
> If you want to increase capacity effectively you MUST be able to use RSS. So verify first if you can use that!

I searched RSS, but now I have 20 people connected to my vpn server and CPU usage has never gone above 10%. According to this, do you think a server with more cores is necessary?

**Reply by Thermi** [reply]
It is entirely load/throughput dependent.

**Reply by meteoguzhan** [reply]
> It is entirely load/throughput dependent.

Thank you very much, I understand very clearly now. 100 people can connect to a 1-core server, but it cannot be used efficiently, but if 100 people connect to a 12-core server, it will be a very efficient use as there will be an average of 9 people per core.

**Reply by Thermi** [reply]
Yes, assuming perfect load balancing of course and a correspondingly bad load profile (if everybody always sends max MTU sized packets then it's probably OK but if everybody sends tiny packets then the throughput will of course be the worst out of all possible scenarios)
